MC/Core/Src/freertos.c

```diff
@@ -0,0 +1,515 @@
+		CAN_set_segment(&send_frame, BATTERY_VOLTAGE_K, mc_data.voltage);
+		CAN_set_segment(&send_frame, BATTERY_CURRENT_K, mc_data.current);
+		CAN_set_segment(&send_frame, MOTOR_TEMP_K, mc_data.motor_temp);
+		CAN_set_segment(&send_frame, MOTOR_CONTROLLER_TEMP_K, mc_data.controller_temp);
+
+		//Send data to RPI
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1)) {
+	    	CAN_send_frame(send_frame);
+	    }
+
+	    osDelay(1);
+
+
+	}
+}
+
+void MC_CommandControlThread(void *argument){
+	CAN_Frame_t command_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER); //initialize frame to receive command from RPI (psuedo frame, RPI code doesn't exist yet)
+	Motor_Controller_t kelly_mc =  MC_init(&hcan3, &hi2c2);; //Initialize motor controller with used CAN and I2C handles
+
+	for(;;){
+
+		//Checks if CAN message is available from RPI
+		if(HAL_CAN_GetRxFifoFillLevel(&hcan3, CAN_RX_FIFO1)){
+			command_frame = CAN_get_frame(&hcan3, CAN_RX_FIFO1);
+		}
+
+		uint8_t command = CAN_get_segment(command_frame, RPI_COMMAND_CODE); //Get "command" from CAN data frame
+
+		if(command == MC_START){
+			MC_drive(&kelly_mc);
+		}
+		else if(command == MC_STOP){
+			MC_stop(&kelly_mc);
+		}
+		else if(command == MC_THROTTLE){
+			uint8_t throttle = CAN_get_segment(command_frame, RPI_COMMAND_DATA); //Get throttle percent "data" from CAN data frame
+			MC_set_throttle(&kelly_mc, (float)throttle);
+		}
+		else if(command == MC_DIRECTION){
+			MC_change_direction();
+		}
+
+	}
+
+}
+
+DrivingDirection bits_to_direction(uint8_t data){
+
+	//Process direction data from MC according to data sheet
+
+	uint8_t data_check = (data & 0b00000011);
+
+	if(data_check == 0b00){
+		return NEUTRAL;
+	}
+	else if(data_check == 0b01){
+		return FORWARD;
+	}
+	else if(data_check == 0b10){
+		return REVERSE;
+	}
+
+	return NEUTRAL;
+}
+
+uint32_t error_code(uint16_t data){
+	return 0;
+}
+
+uint32_t process_bits(uint16_t data){
+	//Processes data from MC according to data sheet (e.g 0.1 C / bit)
+	return data/10;
+}
+
+Motor_Controller_t MC_init(CAN_HandleTypeDef* handler, I2C_HandleTypeDef* i2c_handler) {
+
+	//Initializes Motor_Controller_t struct with base values and proper CAN and I2C handles
+	Motor_Controller_t ret = {
+		.hcan = handler,
+		.direction = 0,
+		.speed = 0,
+		.error_code = 0,
+		.voltage = 0,
+		.current = 0,
+		.motor_temp = 0,
+		.controller_temp = 0,
+		.throttle = DAC_init(i2c_handler)
+	};
+
+	return ret;
+}
+
+void MC_stop(Motor_Controller_t* self) {
+	DAC_write(&self -> throttle, 0); //Sets throttle to 0
+	HAL_GPIO_WritePin(BRK.port, BRK.pin, GPIO_PIN_SET); //Sets brake pin
+}
+
+void MC_drive(Motor_Controller_t* self) {
+	HAL_GPIO_WritePin(BRK.port, BRK.pin, GPIO_PIN_RESET); //Resets brake pin (0)
+	HAL_GPIO_WritePin(REV.port, REV.pin, GPIO_PIN_RESET); //Resets reverse pin
+	HAL_GPIO_WritePin(FW.port, FW.pin, GPIO_PIN_SET); //Sets forward pin
+	DAC_write(&self -> throttle, get_voltage(DEFAULT_THROTTLE)); //Sets throttle to default (100%)
```

Comment:
Maybe not a current concern but isn't full throttle to start kind of a hazard? Thoughts on cutting the default to 50%?

Reply:
It should be fine to leave as is because its super easy to change whenever we want to test. For comp we want to make sure its at 100% right at startup